robot_smach_states/benchmark/benchmark_grasping.py

```diff
@@ -0,0 +1,184 @@
+                    record['observed_class'] = ' '.join([result.type for result in matching_results])
+                else:
+                    record['observed_class'] = matching_results[-1].type
+                selected_entity_id = matching_results[-1].id
+
+                rospy.loginfo("Selected entity {} for grasping".format(selected_entity_id))
+                grasp_entity = ds.EdEntityDesignator(robot, id=selected_entity_id)
+                record['id'] = selected_entity_id[:6]
+
+                entity = grasp_entity.resolve()  # type: Entity
+                if entity:
+                    vector_stamped = entity.pose.extractVectorStamped()  # type: VectorStamped
+                    record['x'] = '{:.3f}'.format(vector_stamped.vector.x())
+                    record['y'] = '{:.3f}'.format(vector_stamped.vector.y())
+                    record['z'] = '{:.3f}'.format(vector_stamped.vector.z())
+
+                grab_state = Grab(robot, grasp_entity, arm)
+
+                record['grab_start'] = time.time()
+                assert grab_state.execute() == 'done', "I couldn't grasp"
+                record['grab_end'] = time.time()
+                record['grab_duration'] = '{:3.3f}'.format(record['grab_end'] - record['grab_start'])
+
+                assert nav_to_start.execute() == 'arrived', "I could not navigate back to the start"
+
+                rospy.logwarn("Robot will turn around to drop the {}".format(grasp_cls.resolve()))
+
+                force_drive = ForceDrive(robot, 0, 0, 1, 3.14)  # rotate 180 degs in pi seconds
+                force_drive.execute()
+
+                say_drop = Say(robot, sentence="I'm going to drop the item, please hold it!")
+                say_drop.execute()
+
+                drop_it = SetGripper(robot, arm_designator=arm, grab_entity_designator=grasp_entity)
+                drop_it.execute()
+
+                force_drive_back = ForceDrive(robot, 0, 0, -1, 3.14)  # rotate -180 degs in pi seconds
+                force_drive_back.execute()
+                nav_to_start.execute()
+            else:
+                raise AssertionError("No {} found".format(grasp_cls.resolve()))
+        else:
+            rospy.logerr("No entities found at all :-(")
+    except AssertionError as assertion_err:
+        say_fail = Say(robot, sentence=assertion_err.message + ", sorry")
+        say_fail.execute()
+    finally:
+        results_writer.writerow(record)
+
+    return record
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(description='Benchmark grasping, for a single item or multiple items at various locations')
+    parser.add_argument("--robot", default="hero",
+                        help="Robot name (amigo, hero, sergio)")
+    parser.add_argument("--output", default="grasp_benchmark.csv",
+                        help="Output of the benchmark results")
+    parser.add_argument("--non-strict-class", action='store_true',
```

Comment:
If you do not care about object recognition you could also just give a '*' to the cls variable right? Is this not double functionality?

Reply:
No, there are several cases I should document more clearly.

- Grab a specific class and fail if that is not available
- Grab a specific class and continue with whatever you find if that is not available.
- Grab any class, of course fail if there is nothing available